controllers/imagepolicy_controller.go

```diff
@@ -172,49 +178,38 @@ func (r *ImagePolicyReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 	}
 
 	if err != nil {
-		imagev1.SetImagePolicyReadiness(
+		conditions.MarkFalse(
 			&pol,
-			metav1.ConditionFalse,
-			meta.ReconciliationFailedReason,
+			meta.ReadyCondition,
+			meta.FailedReason,
 			err.Error(),
 		)
-		if err := r.patchStatus(ctx, req, pol.Status); err != nil {
-			return ctrl.Result{Requeue: true}, err
-		}
-		r.event(ctx, pol, events.EventSeverityError, err.Error())
+		r.Event(ctx, &pol, events.EventSeverityError, EventReasonCannotCalculate, err.Error())
 		return ctrl.Result{}, err
 	}
 
 	if latest == "" {
 		msg := fmt.Sprintf("Cannot determine latest tag for policy: %s", err.Error())
 		pol.Status.LatestImage = ""
-		imagev1.SetImagePolicyReadiness(
+		conditions.MarkFalse(
 			&pol,
-			metav1.ConditionFalse,
-			meta.ReconciliationFailedReason,
+			meta.ReadyCondition,
+			meta.FailedReason,
 			msg,
 		)
-
-		if err := r.patchStatus(ctx, req, pol.Status); err != nil {
-			return ctrl.Result{}, err
-		}
-		r.event(ctx, pol, events.EventSeverityError, msg)
+		r.Event(ctx, &pol, events.EventSeverityError, EventReasonCannotCalculate, msg)
 		return ctrl.Result{}, nil
 	}
 
 	msg := fmt.Sprintf("Latest image tag for '%s' resolved to: %s", repo.Spec.Image, latest)
 	pol.Status.LatestImage = repo.Spec.Image + ":" + latest
-	imagev1.SetImagePolicyReadiness(
+	conditions.MarkTrue(
 		&pol,
-		metav1.ConditionTrue,
-		meta.ReconciliationSucceededReason,
+		meta.ReadyCondition,
+		meta.SucceededReason,
 		msg,
 	)
-
-	if err := r.patchStatus(ctx, req, pol.Status); err != nil {
-		return ctrl.Result{}, err
-	}
-	r.event(ctx, pol, events.EventSeverityInfo, msg)
+	r.Event(ctx, &pol, events.EventSeverityInfo, EventReasonRefFound, msg)
```

Comment:
Users are reporting this is spammy (see https://github.com/fluxcd/notification-controller/issues/234#issuecomment-904200985). I think we should send an event only if a new image tag has been found. The same way source-controller/helm-controller don't send events on no-op reconciliations (same git revision, chart revision).

Reply:
Seems reasonable; let's do it in another PR though, it's not trivial and I want to strictly limit the scope of this one.